lib/stdlib/src/uri_string.erl

```diff
@@ -0,0 +1,2067 @@
+        false -> classify_host_ipv4(Addr)
+    end.
+
+classify_host_ipv4(Addr) ->
+    case is_ipv4_address(Addr) of
+        true -> ipv4;
+        false -> classify_host_regname(Addr)
+    end.
+
+classify_host_regname([]) -> regname;
+classify_host_regname([H|T]) ->
+    case is_reg_name(H) of
+        true -> classify_host_regname(T);
+        false -> other
+    end.
+
+is_ipv4_address(Addr) ->
+    case inet:parse_ipv4strict_address(Addr) of
+        {ok, _} -> true;
+        {error, _} -> false
+    end.
+
+is_ipv6_address(Addr) ->
+    case inet:parse_ipv6strict_address(Addr) of
+        {ok, _} -> true;
+        {error, _} -> false
+    end.
+
+bracket_ipv6(Addr) when is_binary(Addr) ->
+    concat(<<$[,Addr/binary>>,<<$]>>);
+bracket_ipv6(Addr) when is_list(Addr) ->
+    [$[|Addr] ++ "]".
+
+
+%%-------------------------------------------------------------------------
+%% Helper funtions for recompose
+%%-------------------------------------------------------------------------
+
+%%-------------------------------------------------------------------------
+%% Checks if input Map has valid combination of fields that can be
+%% recomposed into a URI.
+%%
+%% The implementation is based on a decision tree that fulfills the
+%% following rules:
+%%   - 'path' shall always be present in the input map
+%%       URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]
+%%       hier-part   = "//" authority path-abempty
+%%                      / path-absolute
+%%                      / path-rootless
+%%                      / path-empty
+%%   - 'host' shall be present in the input map when 'path' starts with
+%%     two slashes ("//")
+%%       path          = path-abempty    ; begins with "/" or is empty
+%%                     / path-absolute   ; begins with "/" but not "//"
+%%                     / path-noscheme   ; begins with a non-colon segment
+%%                     / path-rootless   ; begins with a segment
+%%                     / path-empty      ; zero characters
+%%       path-abempty  = *( "/" segment )
+%%       segment       = *pchar
+%%   - 'host' shall be present if userinfo or port is present in input map
+%%       authority   = [ userinfo "@" ] host [ ":" port ]
+%%   - All fields shall be valid (scheme, userinfo, host, port, path, query
+%%     or fragment).
+%%-------------------------------------------------------------------------
+is_valid_map(#{path := Path} = Map) ->
+    case starts_with_two_slash(Path) of
+        true ->
+            is_valid_map_host(Map);
+        false ->
+            case maps:is_key(userinfo, Map) of
+                true ->
+                    is_valid_map_host(Map);
+                false ->
+                    case maps:is_key(port, Map) of
+                        true ->
+                            is_valid_map_host(Map);
+                        false ->
+                            all_fields_valid(Map)
+                    end
+            end
+    end;
```

Comment:
To avoid case nestling you could do
```
is_valid_map(#{path := Path} = Map) ->
    ((starts_with_two_slash(Path) andalso 
      is_valid_map_host(Map)) 
     orelse
       (maps:is_key(userinfo, Map) andalso 
        is_valid_map_host(Map))  
     orelse
       (maps:is_key(port, Map) andalso 
        is_valid_map_host(Map)) 
     orelse
     all_fields_valid(Map)).
```

Reply:
Fixed.